### ISSUE TITLE
Difference between DRP and Cavern

### ISSUE THREAD

**Issue by kaedewang** [VoidXH/Cavern#91]
contect: dolby Helicopter
render target: 5.1.2 side
drp:
![图片](https://user-images.githubusercontent.com/67850015/221158082-e5093bbe-24ff-4e44-b1ac-65e006c59033.png)

cavern:
![图片](https://user-images.githubusercontent.com/67850015/221158282-ee342474-1aef-4d7d-b680-636b3aa80997.png)

What causes this difference? 
Is the default speaker angle different?

**Comment by VoidXH** [comment]
The single object in this content moves in a circle, which should result in a cosine/sine envelope for the two channels. The result in Cavern is the correct one, DRP is overemphasized. If you set the height channels over 45 degrees of elevation, Cavern can do the same, but I wouldn't recommend it.

**Comment by kaedewang** [comment]
I agree with you. Object moves faster by DRP. Like jumping from one to another.
I also check bird in Amaze. It moves faster, too.
The following is 7.1.4 result. Is it OK in red cycle?
![图片](https://user-images.githubusercontent.com/67850015/221483751-cbaca431-0c02-4aca-a857-65d4ab4356eb.png)

**Comment by VoidXH** [comment]
Yes, the object's speed is not constant. You can take a look in the driver's visualizer.

**Comment by ValZapod** [comment]
You do understand that script for DRP uses DRC, so sound will have smaller amplitude in DRC and 32 bit float is the default in DRC.

**Comment by VoidXH** [comment]
This is not DRC, it doesn't make the sound silent. DRP basically allows this object to move past the room's wall to increase the Atmos effect.

**Comment by ValZapod** [comment]
>basically allows this object to move past the room's wall to increase the Atmos effect

You mean it fakes the coordinates? But the sound occupies much more than a room. It will not be like full football stadium emulation... but with good speakera you can emulate a lot.

**Comment by ValZapod** [comment]
@kaedewang You can use the following command to produce 32 bit integer, not float PCM wav, but on in one channel, but with removed DRC. C:\tmp\atmos\ is where audio is. out-ch-config=13 is 5.1.2 render and audio/x-raw,format=S32LE  is what makes it produce signed 32 bit, not float 32 bit, ac3dec-drc-suppress=true suppresses drc. 32 bit may be compressed into flac, it recently started supporting that.


"C:\Program Files\Dolby\Dolby Reference Player\gst-launch-1.0.exe" --gst-plugin-path "C:\Program Files\Dolby\Dolby Reference Player/gst-plugins" filesrc location=C:\\tmp\\atmos\\gapless.eac3 ! dlbac3parse ! dlbaudiodecbin ac3dec-drc-suppress=true out-ch-config=13 ! audio/x-raw,format=S32LE ! wavenc ! filesink location=C:\\tmp\\atmos\\zE.L.wav

**Comment by kaedewang** [comment]
@ValZapod Max channels of flac is 8. Is it right?
I am consider how to combine 16-16int audio tracks to 8-32int because most players only support 7.1. Audio AMP will separate by TDM setting.
But I don't know how to do it. Do you have some suggestion?

**Comment by kaedewang** [comment]
There is one thing I don't understand.
About DRP render script, why volume are different for different movies?

ex.
Leaf: Volume is similar.
Left DRP; Right Cavern
![图片](https://user-images.githubusercontent.com/67850015/227836745-85ae3ae2-bfe9-4d17-a93c-35f90bbb3479.png)


F1: Cavern is louder.
Left DRP; Right Cavern
![图片](https://user-images.githubusercontent.com/67850015/227836919-375d071e-3a9e-4137-b3a7-3e734506ccaa.png)

**Comment by ValZapod** [comment]
>Max channels of flac is 8. Is it right?

Of course. You will have to separate channels with ffmpeg and use two flac stream. Do NOT use audacity, it is not lossless for wav.

>Left DRP; Right Cavern

Did you use ac3dec-drc-suppress=true that disables DRC?

**Comment by kaedewang** [comment]
> Of course. You will have to separate channels with ffmpeg and use two flac stream. Do NOT use audacity, it is not lossless for wav.
But how to combine TWO-16s track to ONE-32s track? Is it possible?
I want to use 8-ch flac to store 16-ch track.

> Did you use ac3dec-drc-suppress=true that disables DRC?
Disable DRC，Cavern is still louder.
![图片](https://user-images.githubusercontent.com/67850015/227850790-c27ecf33-2f02-47c0-b265-3fd35a3605d5.png)

**Comment by ValZapod** [comment]
>want to use 8-ch flac to store 16-ch track.

Not possible.

>combine TWO-16s track to ONE-32s

You mean bits?

**Comment by kaedewang** [comment]
In 32s, high 16bit stores one channel and low 16bit stores another?

**Comment by ValZapod** [comment]
> In 32s, high 16bit stores one channel and low 16bit stores another?

No. 32 bit integer has each channel 32 bit per sample.

**Comment by VoidXH** [comment]
> Disable DRC, Cavern is still louder.

DRP is always at the -6 dB clipping protection volume, Cavern is only for X.X.2 setups, otherwise it's just -3.

**Comment by kaedewang** [comment]
Is it possible to store wav in eac3? Because kodi supports pass through.
How many channels and how bit depth will pass through for eac3?

**Comment by VoidXH** [comment]
While you can have 16 channels in E-AC-3 with E-AC-3 Merger, it's compressed, mostly in 16-bit precision.

**Comment by kaedewang** [comment]
> E-AC-3 Merger

What's it? Is it a tool? Where can I find?

**Comment by kaedewang** [comment]
Amazing, do you already have used passing through to transform multi-channel wav?

**Comment by VoidXH** [comment]
No, it was released just yesterday.

**Comment by kaedewang** [comment]
How coincidence. So
Is it possible to psss through merged E-AC3?
Can AMP read passed through E-AC3 like wav(tdm)?
My system
Android -USB- UAC -TDM- AMP

**Comment by VoidXH** [comment]
If both the player and DAC supports 7.1, it should work. It's using the same method.

**Comment by kaedewang** [comment]
I try to use merger.
L: 1.wav->5 sample 1111
R: 2.wav->5 sample 2222
![图片](https://user-images.githubusercontent.com/67850015/229181306-f000dce4-bd16-472c-b9aa-7d14132d8323.png)
[12.zip](https://github.com/VoidXH/Cavern/files/11124163/12.zip)
Does it finish merging?

**Comment by VoidXH** [comment]
Added an error check in the new version, but no, stereo is not supported, only 4.0, 5.0, 5.1, or beyond.

**Comment by chris027** [comment]
Is it possible to enable a GUI option in Cavernize to adjust the Listener.EnvironmentSize? I'd like to easily match the DRP setting.

**Comment by VoidXH** [comment]
I checked and unfortunately it's not that simple. The scaling to `EnvironmentSize` is hardcoded to the renderers, making it very hard to _properly_ change. Yes, hacks are possible, you can just rewrite the references locally, but this repository must only contain production-ready code. Making a separate environment scaling would take a longer time.

**Comment by chris027** [comment]
> I checked and unfortunately it's not that simple. The scaling to `EnvironmentSize` is hardcoded to the renderers, making it very hard to _properly_ change. Yes, hacks are possible, you can just rewrite the references locally, but this repository must only contain production-ready code. Making a separate environment scaling would take a longer time.

Thanks so much for looking into this.  I would love to use this app rather than DRP. It’s pretty neat. 

Question: As almost all MLP content over 8 channels (7.1.4 etc…) is Dolby Atmos, why not match DRP output as that’s what Dolby believes it should be and that’s what all mixing engineers are listening to when testing mixes on consumer equipment and DRP?

**Comment by VoidXH** [comment]
@chris027 I found a workaround. This behavior only affects height channels, because DRP doesn't put them on the sides. So if you create a Cavern Driver environment with the height channels not fully at the sides, but closer to the middle (like at 60 degrees vertical), it will do the same. The reason I come to this is that exaggerating wides would put them at the same location as fronts, so we can't do that on the ground level. Keep in mind you need to put the channels at proper Atmos positions to mimic DRP - fronts and rears in the corners, sides exactly to the sides.

**Comment by VoidXH** [comment]
> I wonder why DRP uses a less precise render.

Probably because the reference layouts don't put the top speakers to the sides of the room.

**Comment by chris027** [comment]
> Check the location of the object in Cavern Driver's object viewer. If the object is not perfectly in the corner of the room, that's just proof of yesterday's idea that DRP moves height channels to the middle of the room.

DRP plays the guitar in the top front left channel only, of a 7.1.4 system. 

Cavernize places this in all four height speakers. 

Doesn’t this seem like Cavernize is moving the sound to the middle of the room, not DRP?

**Comment by VoidXH** [comment]
If that's a static mix, please use Cavernize's metadata viewer to get the position of that object and I'll update the layouts accordingly.

**Comment by VoidXH** [comment]
Implemented a workaround in 5ea1d969faadade083b4d1a8844d1d0ee3a18e35.
